docs/guides/validation-user-guide.md

```diff
@@ -4,4 +4,437 @@ headline: Validation User Guide
+
+```proto
+import "spine/options.proto";
+
+message PersonName {
+    option (required_field) = "given_name|honorific_prefix & family_name";
+
+    string honorific_prefix = 1;
+    string given_name = 2;
+    string middle_name = 3;
+    string family_name = 4;
+    string honorific_suffix = 5;
+}
+``` 
+
+In case of `PersonName`, either `given_name` or both `honorific_prefix` and `family_name` must be
+set. All three can be set at the same time.
+
+#### Missing fields
+
+In case if a required field is missing, the validation error message will explicitly say so.
+However, if you need a specific error message for this field, you can provide it via
+the `(if_missing)` option:
+
+```proto
+import "spine/options.proto";
+
+message PhoneNumber {
+    string digits = 1 [(required) = true,
+                       (msg_format = "Phone number must contain digits.")];
+}
+```
+
+Note that this option only applies to fields marked with `(required)` and not to the fields
+referenced via any other options.
+
+If `(goes)` option is used, the error message can be customized with the `(goes).msg_format`
+parameter. Note that the message should contain two "`%s`" insertion points: first for the name of
+the field declaring the option and second for the name of the field targeted by the option.
+
+#### When `(required)` is implicit
+
+In Spine, by convention, if the first declared field of a [Command](../introduction/naming-conventions.html#command-definitions),
+an [Event](../introduction/naming-conventions.html#event-definitions), or an entity state is
+not a collection field, it is considered an entity ID and made required by default. To override this
+convention, use `(required) = false`. 
+
+Example:
+
+```proto
+import "spine/options.proto";
+
+message User {
+    option (entity).kind = AGGREGATE;
+
+    UserId id = 2;
+    PersonName name = 1;
+    // ...
+}
+```
+
+In this case, the `User.id` field is implicitly `(required) = true`. Note that the field __number__
+has nothing to do with this convention, only the field __order__. Thus, `User.name` is not required.
+
+For the next example, consider `user_events.proto`:
+
+```proto
+import "spine/options.proto";
+
+message ProfilePictureChanged {
+
+    Url new_picture = 1 [(required) = false];
+    UserId id = 2;
+}
+```
+
+In this case, the `ProfilePictureChanged.id` field is not required, since it's not declared first
+in the field. The field `ProfilePictureChanged.new_picture` is not required because the convention
+is overridden with an explicit option.
+
+### Recursive message validation
+
+When a message is validated, only the "shallow" constraints are checked by default. This means that
+the message fields can be invalid and the container message is still considered valid.
+
+In order to enable message field checks, use `(validate)` option:
+
+```proto
+import "spine/options.proto";
+
+message User {
+    
+    PersonName name = 2 [(validate) = true];
+    // ...
+}
+```
```

Comment:
We will need more usage examples, for now in Java. We could later have tab splits by platform, like Google does in their docs, — when the time comes.

They are much more illustrative along with the explanation, than the explanation alone.